text/0113-canonical-metadata.md

```diff
@@ -0,0 +1,223 @@
+---
+title: Canonical specification of metadata for LWC modules
+status: DRAFT
+created_at: 2020-05-07
+updated_at: 2020-05-07
+rfc: 
+champion: Aliaksandr Papko (@apapko) | Ravi Jayaramappa (@ravijayaramappa)
+implementation:
+---
+
+# Canonical specification of metadata for LWC modules
+
+Table of Content:
+* [Summary](#summary)
+* [Goals](#goals)
+* [Use Cases](#use-cases)
+* [Current State of Art](#current-state-of-art)
+* [Detailed design](#detailed-design)
+* [References](#reference)
+
+# Summary
+> The [dictionary definition](https://www.merriam-webster.com/dictionary/canonical%20form) of **Canonical form** is `the simplest form of something`. 
+
+This RFC aims to document the metadata collected by statically analyzing the source code of Lightning Web Component(LWC) modules. It documents the standardized shape of metadata.
+
+# Goals
+
+- Completeness
+- Eliminate redundancy: There should be only one way to get a desired information about a module.
+- Consitency in property names
+- Flexibility: Allow for new data to be collected in the future.
+
+# Use Cases
+## Referential Integrity
+One of the key features of the Salesforce platform is referential integrity capabilities. Referential integrity is how the platform allows a given resource to be safely updated and deleted without adverse cascading effects.
+
+## Editor support
+Metadata about components can be used augment standard developer experience in code editors with code completion, attribute name & type validation, peek definition.
+
+## Documentation
+LWC components Metadata can be used to automatically generate documentation about components. Standard javascript documentation formats like JSDoc provided in component can also be gathered as metadata.
+
+## Dependency analysis and prefetching
+Components depend on static resources such as images, javascript files, css files and component definitions to name a few. For offline usage, knowing these dependencies at compile time allows the dependencies to be prefetched and avoid a network call.
+
+## Code generation
+Finally, metadata can facilitate low code development environment by allowing developer to declaratively define component. This metadata can be used to generate component stubs and the developer can add the specific business logic into the stubs thus reducing the amount of manually authored boiler plate code.
```

Comment:
Would be nice to have a proper java version of this to allow us to code-gen components on the server.